README.md
<!-- Copyright 2015-2023, pEp foundation, Switzerland
This file is part of the pEp Engine
This file may be used under the terms of the Creative Commons Attribution-ShareAlike 3.0 Unported (CC BY-SA 3.0) License
See CC_BY-SA.txt -->

![CI Status](https://pep-security.lu/gitlab/cid/mirrors/pepengine/badges/master/pipeline.svg?ignore_skipped=true)

# What is the p≡p Engine?
The p≡p Engine is a Free Software library encapsulating implementations of:

- Key Management

  The p≡p Engine automatically manages keys using a pluggable crypto backend.  The currently implemented backend is based on [Sequoia PGP](https://sequoia-pgp.org), and uses OpenPGP-format keys.

  p≡p for Email is interoperable with OpenPGP.

- Trust Rating

  The p≡p Engine sports a two phase trust rating system for messages:
  In phase one a rating is derived based on channel, crypto and key security.
  This is named "comm_types".
  In phase two these ratings are mapped to user-representable values mimicking the semantics of a traffic light.

- Abstract Crypto API

  The Abstract Crypto API provides functions to encrypt and decrypt data or full messages, without requiring an application programmer to understand the different applied formats and standards.

- Message Transports

  The p≡p Engine will sport a growing list of message transports, to support any widespread text-based messaging system such as E-Mail, SMS, XMPP and many more.

The p≡p Engine is written in C17 and is expected to run on any platform that conforms to the SUS/POSIX specification.
Selected non-SUS platforms are supported as well (such as Microsoft Windows).

# How can I use the p≡p Engine?
The official build instructions can be found at the URL
  https://dev.pep.foundation/Adapter/Adapter%20Build%20Instructions_Version_3.x_DRAFT
The p≡p Engine is not meant to be used in application code directly.
Instead, the p≡p Engine is meant to be used in conjunction with a so-called "adapter".
An adapter provides an API in a programming language that best accomodates developers in their respective software development ecosystem.
So, for example, a Java Native Interface adapter exists for development of apps for the Android mobile operating system, or a .NET adapter exists for development of applications on Microsoft Windows.
Various adapters are also available at the link provided for the p≡p Engine's code above.

# What is the current state of the project?
The p≡p Engine is production-ready.
It is under active development by several full-time employees of the p≡p foundation and its partner organizations.
The most recent version of the source code can be obtained here: <https://gitea.pep.foundation/pEp.foundation/>.
This is the only offical way to obtain a copy of the source code.

# I would like to contribute to the p≡p Engine or a related project, where do I start?
First of all, excellent! You can find further information here: <https://contribution.pep.foundation/>

# Legal notes
The p≡p Engine is Copyright 2015-2023 by p≡p foundation, Switzerland.
The source code of the p≡p Engine is licensed under the terms of the GNU General Public License version 3.
Accompanying documentation is licensed under the terms of the Creative Commons Attribution-ShareAlike 3.0 Unported (CC BY-SA 3.0) License.
Each file includes a notice near its beginning, that indicates the applicable license.
If you wish to license the p≡p Engine under different terms, please contact <mailto:council@pep.foundation>.

src/pEp_rmd160.c is adapted from LibTomCrypt by Tom St Denis and contributors, which was released by its authors into the public domain.

# Contact
The p≡p foundation and the developers of the p≡p Engine can be reached as detailed here: <https://pep.foundation/contact-us/index.html>.


NEWS
v3.3.0-RC5  2023-06-12
* append_string: fix potentially serious pointer bug in utility function.

v3.3.0-RC4  2023-05-26
* Dependencies: use the latest YML, 2.7.6.
* Fix https://gitea.pep.foundation/pEp.foundation/pEpEngine/issues/168 , cleanly.
  Remove the temporary kludge hiding the known violated requirement.

v3.3.0-RC3  2023-05-22
* fix NEWS.
* build system: fix the '-dirty' suffix in the commit identifier generated by
  Makefile.conf .
* As a temporary workaround for
  https://gitea.pep.foundation/pEp.foundation/pEpEngine/issues/168
  replace an assertion with a visible log message printing “WRONG WRONG WRONG
  WRONG”.
  This will make it possible to test the Engine without being bothered by crashes
* DEPENDENCIES: Require pEpEngineSequoiaBackend v1.1.0 , which contains some
  security fixes.

v3.3.0-RC2  2023-05-18
* Make PEP_LOG_ASYNC actually take effect, but calling
  config_enable_log_synchronous *after* the log subsystem has been initialised.
  No visible API change, but when defining PEP_LOG_ASYNC and with the database
  log destination configured in the user should feel the Engine to be faster.
* pEp_abort_unless_PEP_NOABORT: add a session parameter (as the only parameter);
  this is important in case the function is called by adapters.
* build system improvements: add the CPP macros
  * PEP_ENGINE_BRANCH
  * PEP_ENGINE_COMMIT
  * PEP_ENGINE_COMMIT_DATE
  defined by the makefile directly on the compiler command line, as string
  literals.  Platforms not using make to build the Engine should supply these.
* in case of assertion or requirement failure print the expansion of the macros
  above, to help people reporting bugs.  No API change.
* minor internals overhaul: build system, logging entries, whitespace, some
  minor factoring.
* When initialising database connection run the correct pragma:
    PRAGMA foreign_keys=ON;
  with an "s".  The incorrect PRAGMA name I had written before, omitting the
  trailing "s", was silently ignored; some SQL statements in the Engine rely on
  CASCADE constraints being enforced, and the missing constraint caused very
  subtle inconsistencies and changes of behaviour, for example breaking
  replace_userid..
  Closes https://gitea.pep.foundation/pEp.foundation/pEpEngine/issues/166 (this
  bug probably did not affect any released version).

v3.3.0-RC1  2023-05-11
* Make string lists and identity lists nicer to use, with new non-destructive
  functions.  These functions are all declared as API functions in order for
  them to be accessible to the adaptors as well.
  New API functions stringlist_cons, stringlist_reversed, identity_list_cons,
  identity_list_cons_copy, identity_list_reversed .
* Improve SQL concurrency.  The current change should be a proper fix to
  https://gitea.pep.foundation/pEp.foundation/pEpEngine/issues/133 ; I was able
  to see that problem again after upgrading to the latest Thunderbird, but now
  even that is solved.
  Solving this required a trick which is conceptually dirty, even if not complex
  in terms of code size.  No external API change.
* Add the pEp-engine-print-log script, useful to print database-destination logs
  in a compact textual format.  This simple bash script is useful for developers
  and possibly service people.
* Be consistently case-insensitive when handling header names and MIME types.
* Merge dirk's two xcode build-system commits from
  https://gitea.pep.foundation/pEp.foundation/pEpEngine/pulls/159 .
* Remove support for _PEP_SQLITE_DEBUG; fail cleanly at compilation time if a
  user tries to build with it.

v3.2.1  2023-03-23
* DEPENDENCIES: require an official release of libpEpTransport.

v3.2.0  2023-03-23
* CI fixes, thanks to dvn:
  Closes https://gitea.pep.foundation/pEp.foundation/pEpEngine/pulls/147 .
* Update the API specification in api/ so that we can machine-generate more
  code.  Not finished yet, but this might help concurrent development such as
  https://gitea.pep.foundation/pEp.foundation/pEpEngine/issues/149 .
* Solve a benign but distracting intermittent test suite failure, that happens
  on the CI machine but not on positron's laptop.
  Closes https://gitea.pep.foundation/pEp.foundation/pEpEngine/issues/150 .

v3.2.0-RC27  2023-03-16
* Fix management database upgrading from versions older than 13; no API changes.
* Fix benign test suite failures: we are back to zero expected failures.

v3.2.0-RC26  2023-03-03
* Merge https://gitea.pep.foundation/pEp.foundation/pEpEngine/issues/139 from
  Dirk: build system updates, after I added a new compilation unit.

v3.2.0-RC25  2023-03-01
* Merge https://gitea.pep.foundation/pEp.foundation/pEpEngine/pulls/138 (build
  fixes for ios, thanks to Dirk).
* New environment variable PEP_LOG_ASYNC: when defined (to any value) make
  database-destination logging asynchronous, which is less reliable in case of
  crashes but faster by several orders of magnitide.
  New API function to control the same feature at configuration level, per
  session:
  DYNAMIC_API void config_enable_log_synchronous(PEP_SESSION session, bool enable);
* Fix API function get_crashdump_log: it is now back to work correctly if
  logging is enabled with a database destination; otherwise, return a status
  different from PEP_STATUS_OK.
  Closes https://gitea.pep.foundation/pEp.foundation/pEpEngine/issues/91 .
* Update YML2 dependency (reported by Dvn).
* Release builds of the Engine are now compiled (on platforms using make)
  without debugging symbols, in order to work around
  https://gitea.pep.foundation/pEp.foundation/pEpEngine/issues/115 .

v3.2.0-RC24  2023-02-20
* log the current thread id along with the current process id.  When stuck on
  a backoff loop write a log line once every PEP_BACKOFF_TIMES_BEFORE_LOGGING

src/log.h
/**
 * @file    log.h
 * @brief   pEp logging facility
 * @license GNU General Public License 3.0 - see LICENSE.txt
 */

// (setq show-trailing-whitespace t indicate-empty-lines t)

#ifndef PEP_LOG_H
#define PEP_LOG_H

#include "pEpEngine.h"

#include <stdbool.h>
#include <stdio.h>

#ifdef __cplusplus
extern "C" {
#endif

/* Debugging definitions for myself
 * ***************************************************************** */

#define PEP_LOG_LEVEL_MAXIMUM  PEP_LOG_LEVEL_EVERYTHING // FIXME: a test for myself, of course
#define PEP_LOG_DESTINATIONS   (PEP_LOG_DESTINATION_STDERR | PEP_LOG_DESTINATION_DATABASE)


/* Introduction
 * ***************************************************************** */

/* This general-purpose log facility is used internally in the Engine, but is
   also meant for adapters and applications.  The pEp logging facility can
   output information to any number of "destinations" among the supported ones:
   databases, standard error, syslog, system-dependent outputs, and so on.

   Each log entry consists of:
   * a "level", describing the verbosity associated to the current entry;
   * a system name;
   * a sub-system name;
   * a user-specified string;
   * precise source location: source file name, line number, function name;
   * timing information.

   Global configuration allows to suppress long entries above a given verbosity.

   At least when called from C this facility is meant to be used from the macro
   API, which is much more convenient: in particular the macro API allows the
   user to specify the entry string as a literal string template along with a
   set of values to be filled in the template, following the formatted output
   conventions of printf.

   The function API below serves as the internal implementation of the macro
   API, and possibly for calling the log facility from non-C languages. */

/* A convenient way of using this facility for debugging prints is to define
   a CPP macro with a short name and only variadic parameters directly in the
   C file where it will be used; then one can add calls to the local macro in
   the functions being debugged.

   For example imagine a "frob" module in an unnamed application [FIXME:
   Shall I make the system name definable globally?] in frob.c:

   98   #define LOG(...) PEP_LOG_TRACE(NULL, "frob", __VA_ARGS__)
   99
   100  void foo(int x) {
   101      LOG("About to call bar.  x is %i", x);
   102      bar();
   103      int y = x + 2;
   104      LOG("y is %i", y);
   105      LOG();
   106  }
   107
   108  void bar(void) {
   109      LOG("Hello");
   110  }

   Notice the call at line 105, which will only show the function name and
   source location.  The that the template is completely absent in that case:
   one does not even need to supply an empty string.

   Assuming PEP_LOG_LEVEL_MAXIMUM to be at least PEP_LOG_LEVEL_TRACE the log
   will contain something like:

   2022-09-23 20:39:13 frob Trace frob.c:101:foo About to call bar.  x is 10
   2022-09-23 20:39:13 frob Trace frob.c:109:bar Hello
   2022-09-23 20:39:13 frob Trace frob.c:104:foo y is 12
   2022-09-23 20:39:13 frob Trace frob.c:105:foo
*/


/* Parameters
 * ***************************************************************** */

/**
 *  @enum    PEP_LOG_LEVEL
 *
 *  @brief   This can be thought of the verbosity or redundance or
 *           non-importance or a log entry: the higher the value is, the less
 *           important the entry is.
 *
 *           Enabling logging at a certain level means that entries with more
 *           verbosity than the level are ignored.
 */
typedef enum {
    /* A strict lower limit: not intended for actual log entries. */
    PEP_LOG_LEVEL_NOTHING     =   0,

    /* Errors. */
    PEP_LOG_LEVEL_CRITICAL    =  10,
    PEP_LOG_LEVEL_ERROR       =  20,

    /* Warnings. */
    PEP_LOG_LEVEL_WARNING     = 100,
    PEP_LOG_LEVEL_BASIC       = PEP_LOG_LEVEL_WARNING,

    /* Events. */
    PEP_LOG_LEVEL_EVENT       = 200,
    PEP_LOG_LEVEL_API         = 210,
    PEP_LOG_LEVEL_SERVICE     = PEP_LOG_LEVEL_API,

    /* Debugging. */
    PEP_LOG_LEVEL_DEBUG       =  300,
    PEP_LOG_LEVEL_TRACE       =  310,

    /* A strict upper limit: not intended for actual log entries. */
    PEP_LOG_LEVEL_EVERYTHING  = 1000
} PEP_LOG_LEVEL;

/**
 *  @enum    PEP_LOG_DESTINATION_ENUM
 *
 *  @brief   Where to output log entries.  A bitwise-or combination of the
 *           enum values can be used: one copy will be printed to each
 *           destination.
 *           Not every platform will support every possibility; entries logged
 *           to an unsupported destination will simply be discarded.
 */
typedef enum {
    /* Discard log entries. */
    PEP_LOG_DESTINATION_NONE      =  0,

    /* Print log entries in text format to the standard output or the standard
       error stream. */
    PEP_LOG_DESTINATION_STDOUT    =  1,
    PEP_LOG_DESTINATION_STDERR    =  2,

    /* Send log entries to the syslog service. */
    PEP_LOG_DESTINATION_SYSLOG    =  4,

    /* Use the Android logging system. */
    PEP_LOG_DESTINATION_ANDROID   =  8,

    /* Add entries to a local database. */
    PEP_LOG_DESTINATION_DATABASE  = 16,
} PEP_LOG_DESTINATION_ENUM;


/* Configuration
 * ***************************************************************** */

/**
 *  @macro   PEP_LOG_LEVEL_MAXIMUM
 *
 *  @brief   A constant expression of type PEP_LOG_LEVEL, evaluating to
 *           the maximum level of verbosity which is not ignored.
 *
 *           This macro, defined either here or from the compilation
 *           command line, is an upper bound to the log verbosity.  Any
 *           log entry strictly more verbose than this value will not be
 *           printed,  and may be entirely optimised away at compile time.
 */
#if ! defined(PEP_LOG_LEVEL_MAXIMUM)
    /* If the macro has not been defined already (likely from the build system)
       provide a default for it. */
#   if defined(_PEP_SERVICE_LOG_OFF)
#       define PEP_LOG_LEVEL_MAXIMUM  PEP_LOG_LEVEL_NOTHING
#   elif defined(NDEBUG)
#       define PEP_LOG_LEVEL_MAXIMUM  PEP_LOG_LEVEL_SERVICE
#   else
#       define PEP_LOG_LEVEL_MAXIMUM  PEP_LOG_LEVEL_EVERYTHING
#   endif
#endif

/**
 *  @macro   PEP_LOG_DESTINATIONS
 *
 *  @brief   The destinations used for all logging, as a constant
 *           unsigned-integer expression.  For every PEP_LOG_DESTINATION_ENUM
 *           case, if the case bit value is contained in the constant then the
 *           destination is used.
 */
#if ! defined(PEP_LOG_DESTINATIONS)
    /* If the macro has not been defined already (likely from the build system)
       provide a default for it. */
#   if defined(NDEBUG)
#       define PEP_LOG_DESTINATIONS   PEP_LOG_LEVEL_DATABASE
#   else
#       define PEP_LOG_DESTINATIONS   (PEP_LOG_DESTINATION_STDERR       \
                                       | PEP_LOG_DESTINATION_DATABASE)
#   endif
#endif


/* Logging an entry: user macros
 * ***************************************************************** */

/**
 *  <!--       PEP_LOG_CRITICAL()       -->
 *
 *  @brief This macro expands to a C statement logging a message using the
 *         current session, found by capturing a variable named "session"
 *         (assumed to be of type PEP_SESSION) in the current scope.
 *
 *  @param[in]     ??? FIXME: document every parameter: this is important
 *                 ??? The template must be a literal string if supplied, but can be omitted altogether
 *                 ???????????????????????
 *
 */
#define PEP_LOG_CRITICAL(first, ...)                                  \
    PEP_LOG_WITH_LEVEL(PEP_LOG_LEVEL_CRITICAL, (first), __VA_ARGS__)

/**
 *  <!--       PEP_LOG_ERROR()       -->
 *  @brief Exactly like PEP_LOG_CRITICAL, with a different log level.
 */
#define PEP_LOG_ERROR(first, ...)                                  \
    PEP_LOG_WITH_LEVEL(PEP_LOG_LEVEL_ERROR, (first), __VA_ARGS__)

/**
 *  <!--       PEP_LOG_WARNING()       -->
 *  @brief Exactly like PEP_LOG_CRITICAL, with a different log level.
 */
#define PEP_LOG_WARNING(first, ...)                                  \
    PEP_LOG_WITH_LEVEL(PEP_LOG_LEVEL_WARNING, (first), __VA_ARGS__)

/**
 *  <!--       PEP_LOG_EVENT()       -->
 *  @brief Exactly like PEP_LOG_CRITICAL, with a different log level.
 */
#define PEP_LOG_EVENT(first, ...)                                  \
    PEP_LOG_WITH_LEVEL(PEP_LOG_LEVEL_EVENT, (first), __VA_ARGS__)

/**
 *  <!--       PEP_LOG_API()       -->
 *  @brief Exactly like PEP_LOG_CRITICAL, with a different log level.
 */
#define PEP_LOG_API(first, ...)                                  \
    PEP_LOG_WITH_LEVEL(PEP_LOG_LEVEL_API, (first), __VA_ARGS__)

/**
 *  <!--       PEP_LOG_DEBUG()       -->
 *  @brief Exactly like PEP_LOG_CRITICAL, with a different log level.
 */
#define PEP_LOG_DEBUG(first, ...)                                      \
    PEP_LOG_WITH_LEVEL(PEP_LOG_LEVEL_DEBUG, (first), __VA_ARGS__)

/**
 *  <!--       PEP_LOG_TRACE()       -->
 *  @brief Exactly like PEP_LOG_CRITICAL, with a different log level.
 */
#define PEP_LOG_TRACE(first, ...)                                      \
    PEP_LOG_WITH_LEVEL(PEP_LOG_LEVEL_TRACE, (first), __VA_ARGS__)


/* Logging facility: internal macros
 * ***************************************************************** */

/* The macros here are used internally to implement the user macros above. */

/**
 *  <!--       PEP_LOG_WITH_LEVEL()       -->
 *
 * @brief Exactly like PEP_LOG_CRITICAL, with one more parameter prepended
 *        to the others.
 *        Same caveat about capturing a variable named "session".
 *
 *  @param[in]     level       log level, of type PEP_LOG_LEVEL
 *  @param[in]     ...         the remaining parameters are exactly the same
 *                             as in PEP_LOG_CRITICAL.
 *
 *  @warning       This macro is not the most convenient to use from C; the
 *                 macros defined above such as PEP_LOG_CRITICAL and
 *                 PEP_LOG_EVENT are more natural to write in most
 *                 circumstances.
 */
#define PEP_LOG_WITH_LEVEL(first, ...)                                        \
    do {                                                                      \
       /* Capture the local variable named "session" -- static-scoping        \
          purists will not like this -- then call a less convenient macro     \
          doing the actual job.  The first parameter level is not in any way  \
          special, except for its position: defining a variadic macro with    \
          *zero* or more arguments in portable C can be annoying, but we can  \
          entirely avoid the complication here since there will be at least   \
          one parameter. */                                                   \
        PEP_SESSION _pEp_log_session = (session);                             \
                                                                              \
        /* Do the actual work. */                                             \
        PEP_LOG_WITH_SESSION_AND_LEVEL(_pEp_log_session, (first),             \
                                       __VA_ARGS__);                          \
    } while (false)

/**
 *  <!--       PEP_LOG_WITH_SESION_AND_LEVEL()       -->
 *
 * @brief Like PEP_LOG_CRITICAL, with two more parameters prepended
 *        to the others.
 *        Differently from PEP_LOG_CRITICAL this does not capture any
 *        variable named "session": the used session needs to be passed
 *        as a parameter.
 *
 *  @param[in]     session     pEp session
 *  @param[in]     level       log level, of type PEP_LOG_LEVEL
 *  @param[in]     ...         the remaining parameters are exactly the same
 *                             as in PEP_LOG_CRITICAL.
 *
 *  @warning       This macro is not the most convenient to use from C; the
 *                 macros defined above such as PEP_LOG_CRITICAL and
 *                 PEP_LOG_EVENT are more natural to write in most
 *                 circumstances.
 */
#define PEP_LOG_WITH_SESSION_AND_LEVEL(session, level, system, subsystem, ...)  \
    do {                                                                        \
        /* The level argument will be a compile-time constant in almost every   \
           conceivable use case.  This is a good opportunity to optimise away   \
           the entire macro call when this entry exceeds the maximum            \
           verbosity. */                                                        \
        PEP_LOG_LEVEL _pEp_log_level = (level);                                 \
        if (_pEp_log_level > PEP_LOG_LEVEL_MAXIMUM)                             \
            break;                                                              \
                                                                                \
        char *_pEp_log_heap_string;                                             \
        char *_pEp_log_entry;                                                   \
        int _pEp_log_vasprintf_result                                           \
          = pEp_asprintf(& _pEp_log_heap_string,                                \
                         /* Here the beginning of the expansion of              \
                            __VA_ARGS__, if not empty, will be a literal        \
                            string, which will concatenate with "" with no      \
                            harm; but if on the other hand __VA_ARGS__ expands  \
                            to nothing there will still be a valid template.    \
                            This trick makes it possible to use the logging     \
                            facility to trace a line number without supplying   \
                            any explicit string to print at all, not even       \
                            "". */                                              \
                         "" __VA_ARGS__);                                       \
        if (_pEp_log_heap_string == NULL || _pEp_log_vasprintf_result < 0)      \
            /* Allocation failure.  This will be very hard to deal with, but    \
               we can make a desperate attempt of writing some memory           \
               allocation error to the log. */                                  \
            _pEp_log_entry = "could not heap-allocate log string!";             \
        else                                                                    \
            _pEp_log_entry = _pEp_log_heap_string;                              \
        pEp_log((session),                                                      \
                (level),                                                        \
                (system),                                                       \
                (subsystem),                                                    \
                __FILE__,                                                       \
                __LINE__,                                                       \
                __func__,                                                       \
                _pEp_log_entry);                                                \
        free (_pEp_log_heap_string);                                            \
    } while (false)


/* Logging facility: functions
 * ***************************************************************** */

/* PEP_LOG_WITH_SESSION_AND_LEVEL , which despite its inconvenience is the heart
   of every user macro, ultimately relies on the following function.
   The function is not intended for the user, at least in C.  It might be
   convenient for wrapping in other languages. */

/**
 *  <!--       pEp_log()       -->
 *
 *  @brief Emit a log entry to the session destinations in PEP_LOG_DESTINATIONS,
 *         if its level does not pass the verbosity limit PEP_LOG_LEVEL_MAXIMUM.
 *
 *         The log entry time will be the current time.
 *
 *  @param[in]   session          session
 *  @param[in]   level            log level of the new entry
 *  @param[in]   system           the name of the system this log is about;
 *                                this is allowed to be NULL.
 *  @param[in]   subsystem        the name of the subsystem this log is about;
 *                                this is allowed to be NULL.
 *  @param[in]   source_file_name the source file name to log; this is meant
 *                                to be the expansion of __FILE__ at this
 *                                function's call site, but can also be NULL.
 *  @param[in]   source_file_line the source line number to log, 1-based; this
 *                                is meant to be the expansion of __LINE__ at
 *                                this function's call site.
 *  @param[in]   function_name    the function name to log; this is meant
 *                                to be the expansion of __func__ at this
 *                                function's call site, but can also be NULL.
 *  @param[in]   entry            the entry as a human-readable string
 *
 *  @retval PEP_ILLEGAL_VALUE     NULL session
 *  @retval PEP_STATUS_OK         success, if logging to *every* destination
 *                                succeeded
 *  @retval other values          logging to at least one destination failed
 *                                (the function always attempts to log to
 *                                *every* destination)
 *
 */
DYNAMIC_API PEP_STATUS pEp_log(PEP_SESSION session,
                               PEP_LOG_LEVEL level,
                               const char *system,
                               const char *subsystem,
                               const char *source_file_name,
                               int source_file_line,
                               const char *function_name,
                               const char *entry);


/* Logging facility: internal functions implementation
 * ***************************************************************** */

/* A portable implementations of the non-standard GNU/BSD function asprintf
   which performs formatted output on a malloc-allocated string, to be freed by
   the user.  Set * string_pointer to NULL on allocation failure. */
int pEp_asprintf(char **string_pointer, const char *template, ...);


/* Initialisation and finalisation
 * ***************************************************************** */

/* These functions are used internally at the time of session initialisation and
   finalisation.  Not for the user. */
PEP_STATUS pEp_log_initialize(PEP_SESSION session);
PEP_STATUS pEp_log_finalize(PEP_SESSION session);

#ifdef __cplusplus
} /* extern "C" */
#endif

#endif // #ifndef PEP_LOG_H
